### PR TITLE
docs: port the platform integration guides from Discourse

### PR DESCRIPTION
**Pull request by medubelko**
- [x] Have you followed the [guidelines for contributing](https://github.com/canonical/snapcraft/blob/main/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox run -m lint`?
- [x] Have you successfully run `tox run -e test-py310`? (supported versions: `py39`, `py310`, `py311`, `py312`)

-----
CRAFT-3577

### CODE REVIEW

**Comment by medubelko** [comment]
@mr-cal I wanted to give you a preview of the shape, look, and feel of these pages. I aim to have this ready for a full review by end of pulse. If you have high-level concerns, please let me know.

A few things to note:

- The longest file in the stack contains all the unused text and should be ignored. The smaller temp files contain text for the remaining pages.
- Take note that I consider these to flawed how-tos. As we move files away from Discourse, there will be staging fragments of pages that don't have the right home in the quadrants yet. As we adopt more pages and the population starts to be properly partitioned, I'll be correcting these issues and shifting the offending text around.

**Comment by medubelko** [comment]
@mr-cal I wanted to give you a preview of the shape, look, and feel of these pages. I aim to have this ready for a full review by end of pulse. If you have high-level concerns, please let me know.

A few things to note:

- The longest file in the stack contains all the unused text and should be ignored. The smaller temp files contain text for the remaining pages.
- Take note that I consider these to flawed how-tos. As we move files away from Discourse, there will be staging fragments of pages that don't have the right home in the quadrants yet. As we adopt more pages and the population starts to be properly partitioned, I'll be correcting these issues and shifting the offending text around.

**Comment by medubelko** [comment]
@mr-cal I need to separate the example files, but otherwise this should be ready to review.

File: docs/howto/craft-for-platforms/example-compiled-app.rst
Comment (mr-cal) [inline docs/howto/craft-for-platforms/example-compiled-app.rst:3]: From the title, I was expecting a page on C, C++, go, rust, etc platforms.

You use "pre-compiled" in the sentence below and it makes more sense. 

Going further, I like "pre-built" even better because it encapsulates apps using non-compiled languages and aligns terminology with Snapcraft's "build" step.

File: docs/howto/craft-for-platforms/example-python-app.rst
Comment (mr-cal) [inline docs/howto/craft-for-platforms/example-python-app.rst:31]: > Take note that I consider these to flawed how-tos.

I shouldn't be reviewing the content of the how-tos for incorrect stuff like this, right?

File: docs/howto/craft-for-platforms/example-java-app.rst
Comment (mr-cal) [inline docs/howto/craft-for-platforms/example-java-app.rst:15]: Just for me!

File: docs/howto/craft-for-platforms/example-node-app.rst
Comment (mr-cal) [inline docs/howto/craft-for-platforms/example-node-app.rst:86]: I like how you did this

**Review by mr-cal — APPROVED** [review]
Nice work, from a high level the structure and design is great.

File: docs/howto/craft-for-platforms/example-compiled-app.rst
Comment (medubelko) [reply docs/howto/craft-for-platforms/example-compiled-app.rst:3]: Noted, I'll go with "pre-built".

File: docs/howto/craft-for-platforms/example-python-app.rst
Comment (medubelko) [reply docs/howto/craft-for-platforms/example-python-app.rst:31]: If there are technical errors, please point them out!

I'm calling the pages flawed because they're not pure how-tos. The titles aren't exactly appropriate and we're mixing reference material within the text. But given the circumstances they'll serve until we make further strides with the migration.

File: docs/howto/craft-for-platforms/example-java-app.rst
Comment (medubelko) [reply docs/howto/craft-for-platforms/example-java-app.rst:15]: https://en.wikipedia.org/wiki/(Everything_I_Do)_I_Do_It_for_You

File: docs/howto/craft-for-platforms/example-python-app.rst
Comment (mr-cal) [reply docs/howto/craft-for-platforms/example-python-app.rst:31]: Sounds good, I am going to do another review and read the content in detail.

File: docs/howto/craft-for-platforms/index.rst
Comment (medubelko) [reply docs/howto/craft-for-platforms/index.rst]: 2b59176

File: docs/howto/craft-for-platforms/example-compiled-app.rst
Comment (medubelko) [reply docs/howto/craft-for-platforms/example-compiled-app.rst]: 2b59176

File: docs/howto/craft-for-platforms/example-ros-2-app.rst
Comment (mr-cal) [inline docs/howto/craft-for-platforms/example-ros-2-app.rst:314]:
```suggestion
Before launching the app for the first time, you must connect the interfaces.
```

File: docs/howto/craft-for-platforms/example-ros-2-app.rst
Comment (mr-cal) [inline docs/howto/craft-for-platforms/example-ros-2-app.rst:341]: This can be dropped as it's becoming stable this month: https://github.com/canonical/snapcraft/pull/5156

**Review by mr-cal — COMMENTED** [review]
I noticed some plugin reference documentation links point to snapcraft.io and others point to RTD.  Is that intentional?

This is what I am ignoring in this review:
* Some guides use core18 which is technically still supported in snapcraft 7 but newer examples would be preferred. For example, the Rust guide uses core18.
* I am assuming the examples work as they aren't tested and I didn't test them.

**Comment by medubelko** [comment]
@mr-cal 

> I noticed some plugin reference documentation links point to snapcraft.io and others point to RTD. Is that intentional?

Correct. There's no way around this problem at the moment. It's an ugliness we'll have to live with while the migration is in progress.

File: docs/howto/craft-for-platforms/example-dotnet-app.rst
Comment (medubelko) [reply docs/howto/craft-for-platforms/example-dotnet-app.rst:1]: I'm going to leave it out for now, since it's not really relevant to the how-to as-is.

File: docs/howto/craft-for-platforms/example-python-app.rst
Comment (medubelko) [reply docs/howto/craft-for-platforms/example-python-app.rst]: 1684dd99af751120f065d1c021d354e9ab201bf4

File: docs/howto/craft-for-platforms/example-ros-2-app.rst
Comment (medubelko) [reply docs/howto/craft-for-platforms/example-ros-2-app.rst]: 1684dd99af751120f065d1c021d354e9ab201bf4

File: docs/howto/craft-for-platforms/example-ros-2-app.rst
Comment (medubelko) [reply docs/howto/craft-for-platforms/example-ros-2-app.rst]: 1684dd99af751120f065d1c021d354e9ab201bf4

File: docs/howto/craft-for-platforms/example-ros-2-app.rst
Comment (medubelko) [reply docs/howto/craft-for-platforms/example-ros-2-app.rst]: 1684dd99af751120f065d1c021d354e9ab201bf4

**Review by mr-cal — APPROVED** [review]

**Comment by medubelko** [comment]
Merging this will unblock https://github.com/canonical/open-documentation-academy/issues/9.

File: docs/howto/craft-for-platforms/example-c-or-cpp-app.rst
Comment (lengau) [inline docs/howto/craft-for-platforms/example-c-or-cpp-app.rst:46]: Four-space indentation in YAML looks pretty weird to me. I'd suggest two spaces instead, as I think that's what most people are used to seeing and is what [prettier defaults to](https://prettier.io/docs/en/options#tab-width).

```suggestion
    name: moon-buggy
    base: core22
    version: '1.0.51'
    summary: Drive a car across the moon
    description: |
      Moon-buggy is a simple character graphics game, where you drive some
      kind of car across the moon's surface. Unfortunately there are
      dangerous craters there. Fortunately your car can jump over them!
    confinement: devmode

    parts:
      moon-buggy:
        plugin: autotools
        autotools-configure-parameters: ["--prefix=/usr"]
        source: .
        build-packages:
          - libncurses-dev
          - texinfo
    apps:
      moon-buggy:
        command: usr/bin/moon-buggy
```

File: docs/howto/craft-for-platforms/example-python-app.rst
Comment (lengau) [inline docs/howto/craft-for-platforms/example-python-app.rst:63]: If you move the full YAML to its own file (e.g. `example-python-app.yaml`), you can use a literal include for both of these:

```suggestion
  .. literalinclude:: example-python-app.yaml
      :language: yaml

Add a part written in Python
----------------------------
  .. literalinclude:: example-python-app.yaml
      :language: yaml
      :start-at: parts:
      :end-before: apps:
```

**Review by lengau — APPROVED** [review]
The spelling of `wethr` needs fixing, but the other suggestions can wait until later.

File: docs/howto/craft-for-platforms/example-c-or-cpp-app.rst
Comment (medubelko) [reply docs/howto/craft-for-platforms/example-c-or-cpp-app.rst:46]: You're right, I was going to fix this in all the files but got distracted.

File: docs/howto/craft-for-platforms/index.rst
Comment (medubelko) [reply docs/howto/craft-for-platforms/index.rst:12]: Agreed, and I think it will take a while to settle on an alternative.

File: docs/howto/craft-for-platforms/example-c-or-cpp-app.rst
Comment (medubelko) [reply docs/howto/craft-for-platforms/example-c-or-cpp-app.rst]: [342890d](https://github.com/canonical/snapcraft/pull/5160/commits/342890d9e0b2bcb84f35140592fc7c8fe6160b02)

File: docs/howto/craft-for-platforms/example-python-app.rst
Comment (medubelko) [reply docs/howto/craft-for-platforms/example-python-app.rst]: [342890d](https://github.com/canonical/snapcraft/pull/5160/commits/342890d9e0b2bcb84f35140592fc7c8fe6160b02)

**Pull request merged**
